File: backend/src/main/java/com/smartcampus/backend/domain/Notification.java
package com.smartcampus.backend.domain;

import jakarta.persistence.*;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.time.LocalDateTime;

/**
 * In-app notifications delivered to users.
 *
 * Triggered by:
 * - Admin application approved/rejected (→ Campus Admin)
 * - Outlet application approved/rejected (→ Outlet Manager)
 * - Order status change (→ Student) [future]
 */
@JsonIgnoreProperties({ "hibernateLazyInitializer", "handler" })
@Entity
@Table(name = "notifications")
public class Notification {

    // ── Type constants ────────────────────────────────────────────────────────
    public static final String TYPE_ADMIN_APP_APPROVED = "ADMIN_APP_APPROVED";
    public static final String TYPE_ADMIN_APP_REJECTED = "ADMIN_APP_REJECTED";
    public static final String TYPE_OUTLET_APP_APPROVED = "OUTLET_APP_APPROVED";
    public static final String TYPE_OUTLET_APP_REJECTED = "OUTLET_APP_REJECTED";
    public static final String TYPE_ORDER_STATUS_CHANGED = "ORDER_STATUS_CHANGED";
    public static final String TYPE_PENALTY_APPLIED = "PENALTY_APPLIED";
    public static final String TYPE_ORDER_CANCELLED = "ORDER_CANCELLED";
    public static final String TYPE_VERIFICATION_DONE = "VERIFICATION_DONE";

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** The user who receives this notification. */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String message;

    @Column(nullable = false, length = 50)
    private String type;

    @Column(nullable = false)
    private boolean isRead = false;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    protected Notification() {
    }

    public Notification(User user, String title, String message, String type) {
        this.user = user;
        this.title = title;
        this.message = message;
        this.type = type;
        this.createdAt = LocalDateTime.now();
    }

    public Long getId() {
        return id;
    }

    public User getUser() {
        return user;
    }

    public String getTitle() {
        return title;
    }

    public String getMessage() {
        return message;
    }

    public String getType() {
        return type;
    }

    public boolean isRead() {
        return isRead;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void markRead() {
        this.isRead = true;
    }
}

File: backend-fastapi/app/models/notification.py
"""SQLAlchemy model for the `notifications` table. Source: Notification.java"""
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import BigInteger, Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

if TYPE_CHECKING:
    from app.models.user import User


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("users.id"), nullable=False
    )
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    # columnDefinition = "TEXT" in Java
    message: Mapped[str] = mapped_column(Text, nullable=False)
    type: Mapped[str] = mapped_column(String(50), nullable=False)
    # primitive boolean isRead → is_read (SpringPhysicalNamingStrategy)
    is_read: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="notifications", lazy="joined")
